### PR TITLE
Add Application Insights for Spring native

### PR DESCRIPTION
**Pull request by jeanbisutti**
# Description

A project to have Azure  telemetry data with Spring Boot applications and GraalVM native images.

# All SDK Contribution checklist:
- [X] **The pull request does not introduce [breaking changes]**
- [X] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [X] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [X] Title of the pull request is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [X] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Comment by jeanbisutti** [comment]
cc @trask @heyams

**Comment by jeanbisutti** [comment]
The test contained in the `azure-monitor-spring-native-test` module can't be executed today as a GraalVM native executable (`mvn -PnativeTest test`)  beause of #35616 (linked GraalVM issue: https://github.com/oracle/graal/issues/6884).

#35615 was merged into _main_ just after the creation of this PR.

**Comment by lmolkova** [comment]
wow, awesome work @jeanbisutti ! 

I wonder if we want to keep it in the `monitor` folder or move it to `spring` (where all the spring integrations live). And if there should be any specific release schedule or processes we need to coordinate with the spring team (including BOM).

@JonathanGiles @srnagar @stliu @saragluna what do you folks think?

**Comment by jeanbisutti** [comment]
For people outside the Application Insights team, could you please take a special look at the [CI file](https://github.com/Azure/azure-sdk-for-java/pull/35676/files#diff-91a1493e1837140fb34fe37de6ef1a822a817d91a3a171d479282ddb2d391c98) and help with the release build?

**Comment by saragluna** [comment]
Which version of Spring Boot is compatible with `azure-monitor-spring-native`?

**Comment by jeanbisutti** [comment]
> Which version of Spring Boot is compatible with `azure-monitor-spring-native`?

@saragluna It should work from Spring Boot 2. I have not tested with Spring Boot 2.

However, today, the main target of this project is users wanting to have telemetry data on Azure with _Spring Boot 3 native applications_. For the vast majority of other use cases, we recommend using the Java Application Insights agent.

File: sdk/monitor/azure-monitor-spring-native-test/pom.xml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-starter-test</artifactId>
+      <version>3.1.0</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.assertj</groupId>
+      <artifactId>assertj-core</artifactId>
+      <version>3.24.2</version>
```
Comment (trask) [inline sdk/monitor/azure-monitor-spring-native-test/pom.xml:62]: can you check other dependencies also?
```suggestion
      <version>3.22.0</version>  <!-- {x-version-update;org.assertj:assertj-core;external_dependency} -->
```

File: sdk/monitor/azure-monitor-spring-native/src/main/java/com/azure/monitor/applicationinsights/spring/AzureTelemetryActivationConfig.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration(proxyBeanMethods = false)
+public class AzureTelemetryActivationConfig {
+
+    @Bean
+    public AzureTelemetryActivation azureTelemetryActivation(@Value("${applicationinsights.native.spring.non-native.enabled:}") boolean enableEvenWithNonNative) {
```
Comment (trask) [inline sdk/monitor/azure-monitor-spring-native/src/main/java/com/azure/monitor/applicationinsights/spring/AzureTelemetryActivationConfig.java:14]: let's remove this property now that we are going with `azure-monitor-spring-native` for the artifact name, I think the artifact name is specific enough to be clear what the intention is

File: sdk/monitor/azure-monitor-spring-native/src/main/java/com/azure/monitor/applicationinsights/spring/AzureTelemetryConfig.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        LogRecordExporter logRecordExporter = azureMonitorExporterBuilderOpt.get().buildLogRecordExporter();
+        initOTelLogger(logRecordExporter);
+        return logRecordExporter;
+    }
+
+    private void initOTelLogger(LogRecordExporter logRecordExporter) {
+        if (azureMonitorExporterBuilderOpt.isPresent()) {
+            SdkLoggerProvider loggerProvider =
+                    SdkLoggerProvider.builder()
+                            .addLogRecordProcessor(SimpleLogRecordProcessor.create(logRecordExporter))
```
Comment (trask) [inline sdk/monitor/azure-monitor-spring-native/src/main/java/com/azure/monitor/applicationinsights/spring/AzureTelemetryConfig.java:94]: let's use BatchLogRecordProcessor

File: sdk/monitor/azure-monitor-spring-native/src/main/java/com/azure/monitor/applicationinsights/spring/ConnectionStringRetriever.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    private ConnectionStringRetriever() {
+    }
+
+    static Optional<String> retrieveConnectionString(String connectionStringSysProp) {
+
+        if (connectionStringSysProp != null) {
+            if (connectionStringSysProp.isEmpty()) {
+                return Optional.empty();
+            }
+            String connectionStringWithoutDoubleQuotes = connectionStringSysProp.replaceAll("\"", "");
```
Comment (trask) [inline sdk/monitor/azure-monitor-spring-native/src/main/java/com/azure/monitor/applicationinsights/spring/ConnectionStringRetriever.java:23]: i'm not sure about special casing here for connection string only, at least let's add comment to explain/justify the special case

File: sdk/monitor/azure-monitor-spring-native/src/main/resources/META-INF/native-image/native-image.properties [hunk 5]
```diff
@@ -0,0 +1,5 @@
+Args=\
+  --initialize-at-build-time=ch.qos.logback.core \
+  --initialize-at-build-time=ch.qos.logback.classic \
+  --initialize-at-build-time=io.opentelemetry.instrumentation.logback.appender.v1_0.internal.LoggingEventMapper \
+  --initialize-at-build-time=io.opentelemetry.instrumentation.api.internal.cache.concurrentlinkedhashmap.ConcurrentLinkedHashMap
```
Comment (trask) [inline sdk/monitor/azure-monitor-spring-native/src/main/resources/META-INF/native-image/native-image.properties:5]: can you add comments here explaining? (and tracking issue for upstreaming)

File: sdk/monitor/azure-monitor-spring-native/README.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
+# Application Insights for Spring native
+
+This project allows providing telemetry data to Azure Monitor for Spring Boot applications packaged as GraalVM native images.
+
+  
+[Source code][source_code] | [Package (Maven)][package_mvn] | [API reference documentation][api_reference_doc] | [Product Documentation][product_documentation]
```
Comment (trask) [inline sdk/monitor/azure-monitor-spring-native/README.md:6]: looks like first two links are broken

**Review by trask — APPROVED** [review]
💯

**Comment by stliu** [comment]
> wow, awesome work @jeanbisutti !
> 
> I wonder if we want to keep it in the `monitor` folder or move it to `spring` (where all the spring integrations live). And if there should be any specific release schedule or processes we need to coordinate with the spring team (including BOM).
> 
> @JonathanGiles @srnagar @stliu @saragluna what do you folks think?

good point @lmolkova , thanks for bring this up

I think we should keep all our spring integration together, and we're working with Cosmos team and App Configuration team on this, get the spring data cosmos and spring cloud app configuration code relocate into `sdk/spring` folder, so our customer can have a unique palce to find all our spring offerings and build a consistent spring cloud azure project

**Comment by saragluna** [comment]
> > Which version of Spring Boot is compatible with `azure-monitor-spring-native`?
> 
> @saragluna It should work from Spring Boot 2. I have not tested with Spring Boot 2.
> 
> However, today, the main target of this project is users wanting to have telemetry data on Azure with _Spring Boot 3 native applications_. For the vast majority of other use cases, we recommend using the Java Application Insights agent.

Hi @jeanbisutti, we have two active branches for Spring Cloud Azure right now, the `main` branch is for Spring Boot 2 and the `feature/spring-boot-3` branch is for Spring Boot 3. So I am wondering if this project's goal is for Spring Boot 3 native applications, should we put this project in the `feature/spring-boot-3` branch.

File: sdk/monitor/azure-monitor-spring-native/src/main/java/com/azure/monitor/applicationinsights/spring/OtelGlobalRegistrationPostProcessor.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    public OtelGlobalRegistrationPostProcessor(AzureTelemetryActivation azureTelemetryActivation) {
+        this.azureTelemetryActivation = azureTelemetryActivation;
+    }
+
+    @Override
+    public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
+        if (azureTelemetryActivation.isTrue() && bean instanceof OpenTelemetry) {
+            OpenTelemetry openTelemetry = (OpenTelemetry) bean;
+            GlobalOpenTelemetry.set(openTelemetry);
```
[the remark this answers is not in the document]
Comment (jeanbisutti) [reply sdk/monitor/azure-monitor-spring-native/src/main/java/com/azure/monitor/applicationinsights/spring/OtelGlobalRegistrationPostProcessor.java:25]: #35722 created

**Comment by jeanbisutti** [comment]
@lmolkova @stliu @saragluna 

`the main branch is for Spring Boot 2 and the feature/spring-boot-3 branch is for Spring Boot 3. So I am wondering if this project's goal is for Spring Boot 3 native applications, should we put this project in the feature/spring-boot-3 branch.`

The main use case for _Application Insights for Spring native_ is to provide telemetry data for Spring Boot 3 native images. We would also propose this solution for customers with Spring Boot 2 applications and wanting to keep a Java agent other than the Application Insights Java agent.

The version is today 1.0.0-beta.1. So, it is open to modifications as a new location. I will merge the PR not to block additional work. We can continue the discussion about the location of this project after.

**Pull request merged**